### DISCUSSION TITLE
How to use authentication service in app.state instead of global variable

### DISCUSSION THREAD

**Discussion by havardthom** [fastapi/fastapi#3958] [Questions] [answered]
Does anyone here have experience using Authentication service as a part of app.state instead of as a global variable?
This works except openapi/swagger docs wont generate the security scheme. Is there any way around getting security scheme in swagger without passing OAuth2AuthorizationCodeBearerExtended class directly to the Security dependency?

```
async def get_claims(security_scopes: SecurityScopes, request: Request) -> dict:
    request.app.state.auth_service: OAuth2AuthorizationCodeBearerExtended 
    return await request.app.state.auth_service(request, security_scopes)

async def get_current_user(claims: dict = Security(get_claims) -> User:
    print(claims)
    ...
```

**Comment by ghandic** [comment]
Out of curiosity, why don't you want to use the dependency injection?

**Reply by havardthom** [reply]
I am using dependency injection, `get_current_user` is a dependency method. I don't want to use global variables for my services. Service initialization happens on app startup and I think they should be tightly coupled to the FastAPI app itself. E.g. if I want to create two instances of my FastAPI application they would have their own set of services, not a globally shared service. Using a similar pattern to https://github.com/nsidnev/fastapi-realworld-example-app.

The global object vs. app context object has been discussed in multiple issues:
https://github.com/tiangolo/fastapi/issues/81
https://github.com/tiangolo/fastapi/issues/504
https://github.com/tiangolo/fastapi/issues/617

**Reply by ghandic** [reply]
I know you are using it but you were asking for an alternative without saying why 😄 the context helps formulate an answer

**Comment by havardthom** [comment, marked as answer]
Seems like the security scheme only gets set when the passed dependency is of type `SecurityBase` ref. https://github.com/tiangolo/fastapi/blob/fe086a490396eff622a067468531c8776343217f/fastapi/dependencies/utils.py#L151

Doesn't look like there's a way around this. Tried a wrapper method approach without any luck. 
```
async def get_claims_wrapper(request: Request) -> Callable:
    async def get_claims(claims: dict = Security(request.app.state.auth)) -> dict:
        return claims
    return get_claims

async def get_current_user(claims: dict = Depends(get_claims_wrapper)) -> User:
    print(claims)
    ...
```
Seems like my authentication service will have to live as a global object..

Edit:
Watch #3317

**Comment by trzy** [comment]
Facing the same issue in 2024. This seems very poorly architected.
